modules/codegen/src/smithy4s/codegen/IR.scala

```diff
@@ -114,13 +114,18 @@ object Field {
 
 }
 
-case class Alt(name: String, realName: String, tpe: Type, hints: List[Hint])
+case class Alt(
+    name: String,
+    realName: String,
+    tpe: Either[Product, Type],
```

Comment:
I'm definitely open to changing this to a custom ADT if people would prefer.

Reply:
Yeah, let's add a custom ADT. I think it should be a three-way thing : 

``` 
sealed trait UnionMember 
object UnionMember {
   case class EmptyProductCase(...)
   case class ProductCase(product: Product) 
   case class TypeCase(tpe: Type) 
}
```

Reply:
I meant to ask earlier, what is empty product case for? I'm assuming it isn't a product type with zero fields, because that can just be modeled by Product with an empty list of fields, right?

Reply:
Not too sure : when Unit is targeted by a union member we should probably render a case object no matter what, whereas members targeting other empty structures might need a reference to the schema attached to those structures. 

We'll have to give it a think